jocd/src/br/org/certi/jocd/coresight/DebugPort.java

```diff
@@ -0,0 +1,394 @@
+    this.link.setClock(frequency);
+  }
+
+  public void findAps() {
+    int apNum = 0;
+    while (true) {
+      try {
+        long idr = this.readAPNow(((apNum << APSEL_SHIFT) | AP_REG.IDR.getValue()));
+        if (idr == 0) {
+          break;
+        }
+        LOGGER.log(Level.INFO, String.format("AP#%d IDR = 0x%08x", apNum, idr));
+      } catch (Exception e) {
+        LOGGER.log(Level.SEVERE,
+            String.format("Exception reading AP#%d IDR: %s", apNum, e.toString()));
+        break;
+      }
+      apNum += 1;
+    }
+  }
+
+  public long readDPNow(long addr) throws Exception {
+
+    ArrayList<Object> result = readDPLater(addr);
+    Transfer transfer = (Transfer) result.get(0);
+    int num = (int) result.get(1);
+    return readDPAsync(transfer, num);
+  }
+
+  public ArrayList<Object> readDPLater(long addr) throws Exception {
+    assert Reg.containsReg(addr);
+    int num = this.nextAccessNumber();
+
+    Transfer transfer;
+    try {
+      transfer = this.link.readReg(addr);
+      return new ArrayList<>(Arrays.asList(transfer, num));
+    } catch (Error error) {
+      this.handleError(error, num);
+      throw error;
+    }
+  }
+
+  // Read callback returned for async reads.
+  public long readDPAsync(Transfer transfer, int num) throws Exception {
+    try {
+      long result = this.link.readRegAsync(transfer);
+      return result;
+    } catch (Error error) {
+      this.handleError(error, num);
+      throw error;
+    }
+  }
+
+  public boolean writeDP(long addr, long data) throws Error, TimeoutException {
+    assert Reg.containsReg(addr);
+    int num = this.nextAccessNumber();
+
+    // Skip writing DP SELECT register if its value is not changing
+    if (addr == DP_REG.SELECT.getValue()) {
+      if (data == this.dpSelect) {
+        if (LOG_DAP) {
+          LOGGER.log(Level.INFO,
+              String.format("writeDP:%06d cached (addr=0x%08x) = 0x%08x", num, addr, data));
+        }
+        return false;
+      }
+      this.dpSelect = data;
+    }
+
+    // Write the DP register
+    try {
+      if (LOG_DAP) {
+        LOGGER
+            .log(Level.INFO, String.format("writeDP:%06d (addr=0x%08x) = 0x%08x", num, addr, data));
+      }
+      this.link.writeReg(addr, data);
+    } catch (Error error) {
+      this.handleError(error, num);
+      throw error;
+    }
+
+    return true;
+  }
+
+  public boolean writeAP(long addr, Long data) throws TimeoutException, Error {
+    int num = this.nextAccessNumber();
+    long apSel = (addr & APSEL);
+    long bankSel = (addr & APBANKSEL);
+    long apRegaddr = (addr & APREG_MASK);
+
+    // Don't need to write CSW if it's not changing value
+    if (apRegaddr == AP_REG.CSW.getValue()) {
+      if (this.csw.containsKey(apSel) && data == this.csw.get(apSel)) {
+        if (LOG_DAP) {
+          LOGGER.log(Level.INFO,
+              String.format("writeAP:%06d cached (addr=0x%08x) = 0x%08x", num, addr, data));
```

Comment:
I think it is better to make this log message conditional to a certain level instead of a different flag.